mitxgraders/helpers/mitmath/calc.py

```diff
@@ -573,10 +574,39 @@ def handle_node(node):
 
             action = self.actions[node_name]
             handled_kids = [handle_node(k) for k in node]
-            return action(handled_kids)
 
-        # Find the value of the entire tree.
-        result = handle_node(self.tree)
+            # Check for nan
+            if any(np.isnan(item) for item in handled_kids if isinstance(item, float)):
+                return float('nan')
+
+            # Compute the result of this node
+            result = action(handled_kids)
+
+            # All actions convert the input to a number, array or IdentityMultiple
+            # Check if there were any infinities or nan
+            check = result
+            # IdentityMultiple needs to extract the value
+            if isinstance(result, IdentityMultiple):
+                check = result.value
+            if not allow_inf and np.any(np.isinf(check)):
```

Comment:
Currently our factorial function uses `math.factorial` if the input is an integer (or a float that can be interpreted as an integer) and `math.factorial` returns type `<long>` if its input is greater than 20. And `isinf` doesn't work for `<long>`.

Suggestion: scrap `math.factorial` all together, and go straight to the scipy gamma+1 implementation.
- The only difference will be `fact(3)=6.0` instead of `6`. I don't think this matters for anything at all, since MathArrays can handle integer-like powers.
- It would also resolve the issue of `fact(1000000)` timing out edX.

Reply:
Ok, will do.

Reply:
Done.